README.md
Pre Orders in Construction


docs/history.md
# BACKEND

## Create a proyect

@mkdir orders
@cd orders

## Create a Django Proyect with restframework

:ctrl+alt+t
:ctrl+alt+shift+t
> name: back

@python -m venv orders-env
@source orders-env/source/activate
@pip install --upgrade pip
@pip install django djangorestframework
@django-admin startproject back
@cd back
@python manage.py startapp orders

## Prepare nvim 

:ctrl+alt+t
:ctrl+alt+shift+t
> name: nvim

nvim


## Register App

Edit -> orders/back/back/settings.py

'''
INSTALLED_APPS = [
    #...
    #new_apps
    'orders',
    'rest_framework',
]
'''

## Create models in orders

Edit -> orders/back/orders/models.py

-add

from django.contrib.auth.models import User

Create class follow:

* class Color(modelos.Model):
* class Carriage(models.Model):
* class Category(models.Model):
* class Product(models.Model):
* class Client(models.Model):
* class Price_list(models.Model):
* class Line_price_list(models.Model):
* class Order(models.Model):
* class Line_order(models.Model):

## Inser the common fields:


    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='user_customer')
    charge_date = models.DateTimeField(auto_now_add=True)
    name = models.CharField(max_length=100, unique=True)

in the line 1- related_name is "user" + class name.

## Make Migrations and create super user


@python manage.py makemigrations
@python manage.py migrate
@mpoython manage.py createsuperuser

## Implemented JWT

create folder -> orders/back/orders/api
in api create files:
* __init__.py
* views.py
* serializers.py
* urls.py

@pip install djangorestframework-simplejwt

in -> ordes/back/back/setting add:

'''
REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework_simplejwt.authentication.JWTAuthentication',
    )
}
'''

import:

'''

from datetime import timedelta
'''

then:

in INSTALLED_APPS add :

'''
'rest_framework_simplejwt.token_blacklist'
'''



'''
SIMPLE_JWT = {
    "ACCESS_TOKEN_LIFETIME": timedelta(minutes=15),
    "REFRESH_TOKEN_LIFETIME": timedelta(days=90),
    "ROTATE_REFRESH_TOKENS": True,
    "BLACKLIST_AFTER_ROTATION": True,
    "UPDATE_LAST_LOGIN": False,

    "ALGORITHM": "HS256",
    "VERIFYING_KEY": "",
    "AUDIENCE": None,
    "ISSUER": None,
    "JSON_ENCODER": None,
    "JWK_URL": None,
    "LEEWAY": 0,

    "AUTH_HEADER_TYPES": ("Bearer",),
    "AUTH_HEADER_NAME": "HTTP_AUTHORIZATION",
    "USER_ID_FIELD": "id",
    "USER_ID_CLAIM": "user_id",
    "USER_AUTHENTICATION_RULE": "rest_framework_simplejwt.authentication.default_user_authentication_rule",

    "AUTH_TOKEN_CLASSES": ("rest_framework_simplejwt.tokens.AccessToken",),
    "TOKEN_TYPE_CLAIM": "token_type",
    "TOKEN_USER_CLASS": "rest_framework_simplejwt.models.TokenUser",

    "JTI_CLAIM": "jti",

    "SLIDING_TOKEN_REFRESH_EXP_CLAIM": "refresh_exp",
    "SLIDING_TOKEN_LIFETIME": timedelta(minutes=5),
    "SLIDING_TOKEN_REFRESH_LIFETIME": timedelta(days=1),

    "TOKEN_OBTAIN_SERIALIZER": "rest_framework_simplejwt.serializers.TokenObtainPairSerializer",
    "TOKEN_REFRESH_SERIALIZER": "rest_framework_simplejwt.serializers.TokenRefreshSerializer",
    "TOKEN_VERIFY_SERIALIZER": "rest_framework_simplejwt.serializers.TokenVerifySerializer",
    "TOKEN_BLACKLIST_SERIALIZER": "rest_framework_simplejwt.serializers.TokenBlacklistSerializer",
    "SLIDING_TOKEN_OBTAIN_SERIALIZER": "rest_framework_simplejwt.serializers.TokenObtainSlidingSerializer",
    "SLIDING_TOKEN_REFRESH_SERIALIZER": "rest_framework_simplejwt.serializers.TokenRefreshSlidingSerializer",
}
'''
in -> orders/back/orders/api/view.py

'''
from rest_framework_simplejwt.serializers import TokenObtainPairSerializer
from rest_framework_simplejwt.views import TokenObtainPairView

class MyTokenObtainPairSerializer(TokenObtainPairSerializer):
    @classmethod
    def get_token(cls, user):
        token = super().get_token(user)
        token['username'] = user.username
        return token

class MyTokenObtainPairView(TokenObtainPairView):
    serializer_class = MyTokenObtainPairSerializer


'''

in -> orders/back/orders/api/urls.py

'''

from django.urls import path
from .views import MyTokenObtainPairView
from rest_framework_simplejwt.views import (
    TokenRefreshView,
)

urlpatterns = [
        path('token/', MyTokenObtainPairView.as_view(), name='token_obtain_pair'),
        path('token/refresh/', TokenRefreshView.as_view(), name='token_refresh'),
        ]
'''

@python manage.py migrate
@python manage.py runserver

## Configure CORS

pip install django-cors-headers

in -> orders/back/back/settings.py

'''
INSTALLED_APPS = [
...
   'corsheaders',
...
]
'''


'''
MIDDLEWARE = [
    ...
    'corsheaders.middleware.CorsMiddleware',
    ...
]

'''

and add in settings.py

'''
CORS_ALLOW_ALL_ORIGINS = True

'''

Finish backend.

# FRONTEND


## Create a react proyect

:ctrl+alt+t
:ctrl+alt+shift+t
> name: front

@npm create vite@latest front
>React
>Javascript
@cd front
@npm install
@npm run dev

output:

'''
> front@0.0.0 dev
> vite


  VITE v4.1.4  ready in 722 ms

  ➜  Local:   http://localhost:5173/
  ➜  Network: use --host to expose
  ➜  press h to show help
'''


## Clean Proyect... empty

only files:

App.css
App.jsx
main.jsx

and create on sources the follow dirs:

context
pages
components
utils

then, on pages create: HomePage.jsx and LoginPage.jsx

with rafce snipped create booth class


npm install react-router-dom

in main.jsx


import {BrowserRouter} from 'react-router-dom'
...


<React.StrictMode>
   <BrowserRouter>
     <App />
   </BrowserRouter>
</React.StrictMode>,


Create file in folder "components" Header.jsx:

import React from 'react'
import { Link } from 'react-router-dom'
const Header = () => {
  return (
    <div>
      <Link to="/">Home</Link>
      <span> | </span>
      <Link to="/login">Login</Link>
    </div>
  )
}

export default Header


Create file in -> utils

PrivateRouter.jsx

import { Navigate, Outlet } from 'react-router-dom'
import React from 'react'


back/orders/models.py
from django.db import models
from django.contrib.auth.models import User

PROVINCE = [ 
    ('CF', 'CAPITAL FEDERAL'),
    ('BA', 'BUENOS AIRES'),
    ('CA', 'CATAMARCA'),
    ('CH', 'CHACO'),
    ('CU', 'CHUBUT'),
    ('CO', 'CÓRDOBA'),
    ('CR', 'CORRIENTES'),
    ('ER', 'ENTRE RÍOS'),
    ('FO', 'FORMOSA'),
    ('JU', 'JUJUY'),
    ('LP', 'LA PAMPA'),
    ('LR', 'LA RIOJA'),
    ('MZ', 'MENDOZA'),
    ('MI', 'MISIONES'),
    ('NQ', 'NEUQUÉN'),
    ('RN', 'RÍO NEGRO'),
    ('SA', 'SALTA'),
    ('SJ', 'SAN JUAN'),
    ('SL', 'SAN LUIS'),
    ('SC', 'SANTA CRUZ'),
    ('SF', 'SANTA FE'),
    ('SE', 'SANTIAGO DEL ESTERO'),
    ('TF', 'TIERRA DEL FUEGO'),
    ('TU', 'TUCUMÁN')
            ]

BILLING = [ 
        ('1' , 'PRO'),
        ('2', 'FC')
           ]

class Color(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='created_user_color')
    charge_date = models.DateTimeField(auto_now_add=True)
    name = models.CharField(max_length=100, unique=True)
    image = models.CharField(max_length=200, unique=True)

    def __str__(self):
        return self.name


class Carriage(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='created_user_carriage')
    charge_date = models.DateTimeField(auto_now_add=True)
    name = models.CharField(max_length=100, unique=True)
    address = models.CharField(max_length=200)
    location = models.CharField(max_length=100)
    phone = models.CharField(max_length=100, blank=True)
    web = models.CharField(max_length=200, blank=True)
    enlaced_number_system = models.IntegerField()
    enlaced_code_system = models.CharField(max_length=20, blank=True)
    observations = models.TextField(blank=True)

class Category(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='created_user_category')
    charge_date = models.DateTimeField(auto_now_add=True)
    name = models.CharField(max_length=100, unique=True)
    image = models.CharField(max_length=200, blank=True)
    observations = models.TextField()

class Product(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='created_user_product')
    charge_date = models.DateTimeField(auto_now_add=True)
    name = models.CharField(max_length=100, unique=True)
    description = models.TextField()
    materials = models.CharField(max_length=200)
    weight = models.FloatField()
    size = models.CharField(max_length=100)
    catergory = models.ForeignKey(Category, on_delete=models.CASCADE, related_name='category_product')
    units_per_package = models.IntegerField()
    observations = models.TextField()

class Client(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='created_user_client')
    charge_date = models.DateTimeField(auto_now_add=True)
    name = models.CharField(max_length=100, unique=True)
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='user_price_list')
    charge_date = models.DateTimeField(auto_now_add=True)
    phone = models.CharField(max_length=100, blank=True)
    cuit = models.CharField(max_length=20, blank=True)
    address = models.CharField(max_length=200)
    province = models.CharField(max_length=2, choices=PROVINCE, default='BA')
    location = models.CharField(max_length=100)
    observations = models.TextField(blank=True)

class Price_list(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='created_user_price_list')
    charge_date = models.DateTimeField(auto_now_add=True)
    name = models.CharField(max_length=100, unique=True)
    initial_date = models.DateField()
    finish_date = models.DateField()
    observations = models.TextField(blank=True)
    enlaced_number_system = models.IntegerField()
    enlaced_code_system = models.CharField(max_length=30, blank=True)

class Line_price_list(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='created_user_line_price')
    charge_date = models.DateTimeField(auto_now_add=True)
    product = models.ForeignKey(Product, on_delete=models.CASCADE, related_name='product_line_price_list')
    price = models.FloatField()

class Order(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='created_user_order')
    charge_date = models.DateTimeField(auto_now_add=True)
    price_list = models.ForeignKey(Price_list, on_delete=models.CASCADE, related_name='price_list_order')
    client = models.ForeignKey(Client, on_delete=models.CASCADE, related_name='client_order')
    carriage = models.ForeignKey(Carriage, on_delete=models.CASCADE, related_name='carriage_order')
    billing = models.CharField(max_length=1, choices=BILLING, default='2')

class Line_order(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='created_user_line_order')
    charge_date = models.DateTimeField(auto_now_add=True)
    order = models.ForeignKey(Order, on_delete=models.CASCADE, related_name='order_line_order')
    product = models.ForeignKey(Product, on_delete=models.CASCADE, related_name='product_line_order')
    amount = models.IntegerField()
    price = models.FloatField()
    color = models.ForeignKey(Color, on_delete=models.CASCADE, related_name='color_line_order')
    billing = models.CharField(max_length=1, choices=BILLING, default='2')
    discount = models.IntegerField()


back/orders/api/views.py
from rest_framework_simplejwt.serializers import TokenObtainPairSerializer
from rest_framework_simplejwt.views import TokenObtainPairView
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response
from rest_framework.decorators import api_view, permission_classes
from .serializers import ColorSerializer
from orders.models import Color



class MyTokenObtainPairSerializer(TokenObtainPairSerializer):
    @classmethod
    def get_token(cls, user):
        token = super().get_token(user)
        token['username'] = user.username
        return token

class MyTokenObtainPairView(TokenObtainPairView):
    serializer_class = MyTokenObtainPairSerializer


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def get_colors(request):
    #user = request.user   
    #colorx =  user.color.all()
    colors = Color.objects.all()
    serializer = ColorSerializer(colors, many=True)
    return Response(serializer.data)


back/orders/migrations/0001_initial.py
# Generated by Django 4.1.7 on 2023-03-12 03:59

from django.conf import settings
from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name='Carriage',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('charge_date', models.DateTimeField(auto_now_add=True)),
                ('name', models.CharField(max_length=100, unique=True)),
                ('address', models.CharField(max_length=200)),
                ('location', models.CharField(max_length=100)),
                ('phone', models.CharField(blank=True, max_length=100)),
                ('web', models.CharField(blank=True, max_length=200)),
                ('enlaced_number_system', models.IntegerField()),
                ('enlaced_code_system', models.CharField(blank=True, max_length=20)),
                ('observations', models.TextField(blank=True)),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='created_user_carriage', to=settings.AUTH_USER_MODEL)),
            ],
        ),
        migrations.CreateModel(
            name='Category',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('charge_date', models.DateTimeField(auto_now_add=True)),
                ('name', models.CharField(max_length=100, unique=True)),
                ('image', models.CharField(blank=True, max_length=200)),
                ('observations', models.TextField()),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='created_user_category', to=settings.AUTH_USER_MODEL)),
            ],
        ),
        migrations.CreateModel(
            name='Client',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(max_length=100, unique=True)),
                ('charge_date', models.DateTimeField(auto_now_add=True)),
                ('phone', models.CharField(blank=True, max_length=100)),
                ('cuit', models.CharField(blank=True, max_length=20)),
                ('address', models.CharField(max_length=200)),
                ('province', models.CharField(choices=[('CF', 'CAPITAL FEDERAL'), ('BA', 'BUENOS AIRES'), ('CA', 'CATAMARCA'), ('CH', 'CHACO'), ('CU', 'CHUBUT'), ('CO', 'CÓRDOBA'), ('CR', 'CORRIENTES'), ('ER', 'ENTRE RÍOS'), ('FO', 'FORMOSA'), ('JU', 'JUJUY'), ('LP', 'LA PAMPA'), ('LR', 'LA RIOJA'), ('MZ', 'MENDOZA'), ('MI', 'MISIONES'), ('NQ', 'NEUQUÉN'), ('RN', 'RÍO NEGRO'), ('SA', 'SALTA'), ('SJ', 'SAN JUAN'), ('SL', 'SAN LUIS'), ('SC', 'SANTA CRUZ'), ('SF', 'SANTA FE'), ('SE', 'SANTIAGO DEL ESTERO'), ('TF', 'TIERRA DEL FUEGO'), ('TU', 'TUCUMÁN')], default='BA', max_length=2)),
                ('location', models.CharField(max_length=100)),
                ('observations', models.TextField(blank=True)),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='user_price_list', to=settings.AUTH_USER_MODEL)),
            ],
        ),
        migrations.CreateModel(
            name='Color',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('charge_date', models.DateTimeField(auto_now_add=True)),
                ('name', models.CharField(max_length=100, unique=True)),
                ('image', models.CharField(max_length=200, unique=True)),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='created_user_color', to=settings.AUTH_USER_MODEL)),
            ],
        ),
        migrations.CreateModel(
            name='Product',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('charge_date', models.DateTimeField(auto_now_add=True)),
                ('name', models.CharField(max_length=100, unique=True)),
                ('description', models.TextField()),
                ('materials', models.CharField(max_length=200)),
                ('weight', models.FloatField()),
                ('size', models.CharField(max_length=100)),
                ('units_per_package', models.IntegerField()),
                ('observations', models.TextField()),
                ('catergory', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='category_product', to='orders.category')),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='created_user_product', to=settings.AUTH_USER_MODEL)),
            ],
        ),
        migrations.CreateModel(
            name='Price_list',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('charge_date', models.DateTimeField(auto_now_add=True)),
                ('name', models.CharField(max_length=100, unique=True)),
                ('initial_date', models.DateField()),
                ('finish_date', models.DateField()),
                ('observations', models.TextField(blank=True)),
                ('enlaced_number_system', models.IntegerField()),
                ('enlaced_code_system', models.CharField(blank=True, max_length=30)),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='created_user_price_list', to=settings.AUTH_USER_MODEL)),
            ],
        ),
        migrations.CreateModel(
            name='Order',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('charge_date', models.DateTimeField(auto_now_add=True)),
                ('billing', models.CharField(choices=[('1', 'PRO'), ('2', 'FC')], default='2', max_length=1)),
                ('carriage', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='carriage_order', to='orders.carriage')),
                ('client', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='client_order', to='orders.client')),
                ('price_list', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='price_list_order', to='orders.price_list')),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='created_user_order', to=settings.AUTH_USER_MODEL)),
            ],
        ),
        migrations.CreateModel(
            name='Line_price_list',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('charge_date', models.DateTimeField(auto_now_add=True)),
                ('price', models.FloatField()),
                ('product', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='product_line_price_list', to='orders.product')),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='created_user_line_price', to=settings.AUTH_USER_MODEL)),
            ],
        ),
        migrations.CreateModel(
            name='Line_order',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('charge_date', models.DateTimeField(auto_now_add=True)),
                ('amount', models.IntegerField()),
                ('price', models.FloatField()),
                ('billing', models.CharField(choices=[('1', 'PRO'), ('2', 'FC')], default='2', max_length=1)),
                ('discount', models.IntegerField()),
                ('color', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='color_line_order', to='orders.color')),
                ('order', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='order_line_order', to='orders.order')),
                ('product', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='product_line_order', to='orders.product')),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='created_user_line_order', to=settings.AUTH_USER_MODEL)),
            ],
        ),
    ]


back/back/settings.py
"""
Django settings for back project.

Generated by 'django-admin startproject' using Django 4.1.7.

For more information on this file, see
https://docs.djangoproject.com/en/4.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/4.1/ref/settings/
"""

from pathlib import Path

from datetime import timedelta
# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/4.1/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = 'django-insecure-(*s^#&_6==8hxok&k!)^-cu7&li+oil(8dbwje**dt2!-m&4*2'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = ['*']


# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    #new_apps
    'orders',
    'rest_framework',
    'rest_framework_simplejwt.token_blacklist',
    'corsheaders',
]

REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': [
        'rest_framework_simplejwt.authentication.JWTAuthentication',
        ]
}

SIMPLE_JWT = {
    "ACCESS_TOKEN_LIFETIME": timedelta(minutes=15),
    "REFRESH_TOKEN_LIFETIME": timedelta(days=90),
    "ROTATE_REFRESH_TOKENS": True,
    "BLACKLIST_AFTER_ROTATION": True,
    "UPDATE_LAST_LOGIN": False,

    "ALGORITHM": "HS256",
    "VERIFYING_KEY": "",
    "AUDIENCE": None,
    "ISSUER": None,
    "JSON_ENCODER": None,
    "JWK_URL": None,
    "LEEWAY": 0,

    "AUTH_HEADER_TYPES": ("Bearer",),
    "AUTH_HEADER_NAME": "HTTP_AUTHORIZATION",
    "USER_ID_FIELD": "id",
    "USER_ID_CLAIM": "user_id",
    "USER_AUTHENTICATION_RULE": "rest_framework_simplejwt.authentication.default_user_authentication_rule",

    "AUTH_TOKEN_CLASSES": ("rest_framework_simplejwt.tokens.AccessToken",),
    "TOKEN_TYPE_CLAIM": "token_type",
    "TOKEN_USER_CLASS": "rest_framework_simplejwt.models.TokenUser",

    "JTI_CLAIM": "jti",

    "SLIDING_TOKEN_REFRESH_EXP_CLAIM": "refresh_exp",
    "SLIDING_TOKEN_LIFETIME": timedelta(minutes=5),
    "SLIDING_TOKEN_REFRESH_LIFETIME": timedelta(days=90),

    "TOKEN_OBTAIN_SERIALIZER": "rest_framework_simplejwt.serializers.TokenObtainPairSerializer",
    "TOKEN_REFRESH_SERIALIZER": "rest_framework_simplejwt.serializers.TokenRefreshSerializer",
    "TOKEN_VERIFY_SERIALIZER": "rest_framework_simplejwt.serializers.TokenVerifySerializer",
    "TOKEN_BLACKLIST_SERIALIZER": "rest_framework_simplejwt.serializers.TokenBlacklistSerializer",
    "SLIDING_TOKEN_OBTAIN_SERIALIZER": "rest_framework_simplejwt.serializers.TokenObtainSlidingSerializer",
    "SLIDING_TOKEN_REFRESH_SERIALIZER": "rest_framework_simplejwt.serializers.TokenRefreshSlidingSerializer",
}

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'back.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'back.wsgi.application'


# Database
# https://docs.djangoproject.com/en/4.1/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': BASE_DIR / 'db.sqlite3',
    }
}


# Password validation
# https://docs.djangoproject.com/en/4.1/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/4.1/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/4.1/howto/static-files/

STATIC_URL = 'static/'

# Default primary key field type
# https://docs.djangoproject.com/en/4.1/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'

CORS_ALLOW_ALL_ORIGINS = True
CORS_ALLOWED_ORIGINS = ['http://127.0.0.1:5173', 'http://localhost:5173']
